Member password reset

nixpulvis opened the pull request:
Member need to be able to reset their own passwords.

For more information see https://trello.com/c/rDBe1FVZ

nixpulvis committed:
Adding ResetKey class

https://trello.com/c/rDBe1FVZ

To handle a member forgeting his/her password and needing a
key to be emailed, for the purpose of validating ID.

nixpulvis committed:
Adding the member -> reset_key relation

A member has_one reset_key. Never more, but the key may be 'nil'.

nixpulvis committed:
ResetKey in Member integration and testing.

nixpulvis committed:
Adding a comment for the forgot_password! method.

nixpulvis committed:
Proper default value setting for ResetKey's key.

Need to use ||= to avoid overwriting the ResetKey's key
on every call to initialize.

nixpulvis committed:
Adding default value for valid_until

nixpulvis committed:
Don't set the valid_until value in the factory.

nixpulvis committed:
Ensure we can set valid until.

nixpulvis committed:
Protecting ResetKey key value from mass-assignment

Testing these changes with shoulda.

nixpulvis committed:
Use default ResetKey valid_until date.

nixpulvis committed:
Action for creating a ResetKey for members by email.

A POST to 'forgot-password' with an email, will look up the
member by email, and if the member exists, calls forgot_password!
on that member. This method creates the ResetKey, and will email
the user with instructions on using it. The page rendered is a
simple "Email Sent" page.

If the email is not that of a member, then it will redirect back
and flash an error message.

pranav commented:
Why aren't we using some kind of Facebook/Google/CCIS authentication for this?
—
Sent from Mailbox for iPhone

On Wed, May 1, 2013 at 1:09 PM, Nathan Lilienthal
notifications@github.com wrote:

> Member need to be able to reset their own passwords.
> For more information see https://trello.com/c/rDBe1FVZ
> You can merge this Pull Request by running:
>   git pull https://github.com/nuacm/website member-password-reset
> Or you can view, comment on it, or merge it online at:
>   https://github.com/nuacm/website/pull/1
> -- Commit Summary --
> - Adding ResetKey class
> - Adding the member -&gt; reset_key relation
> - ResetKey in Member integration and testing.
> - Adding a comment for the forgot_password! method.
> - Proper default value setting for ResetKey&#39;s key.
> - Adding default value for valid_until
> - Don&#39;t set the valid_until value in the factory.
> - Ensure we can set valid until.
> - Protecting ResetKey key value from mass-assignment
> - Use default ResetKey valid_until date.
> - Action for creating a ResetKey for members by email.
>   -- File Changes --
>   M Gemfile (3)
>   M Gemfile.lock (3)
>   M app/controllers/members_controller.rb (11)
>   M app/models/member.rb (14)
>   A app/models/reset_key.rb (14)
>   A app/views/members/forgot_password.html.erb (1)
>   M config/routes.rb (2)
>   A db/migrate/20130501141520_create_reset_keys.rb (10)
>   A db/migrate/20130501144647_add_member_id_to_reset_key.rb (5)
>   M db/schema.rb (10)
>   M spec/controllers/members_controller_spec.rb (31)
>   A spec/factories/reset_key.rb (4)
>   D spec/helpers/members_helper_spec.rb (15)
>   M spec/models/member_spec.rb (27)
>   A spec/models/reset_key_spec.rb (24)
>   -- Patch Links --
>   https://github.com/nuacm/website/pull/1.patch
>   https://github.com/nuacm/website/pull/1.diff

nixpulvis commented:
See the comment on trello, It's a possibility.

nixpulvis committed:
Adding a few tests and starting time based testing.

Need to integrate timecop to finish this.

nixpulvis committed:
Adding time based testing of ResetKey

Using timecop to alter time and test the validity
of the ResetKeys.

nixpulvis committed:
ResetKey explained in comments.

nixpulvis committed:
Test ResetKey create from POST forgot-password.

nixpulvis committed:
Merge branch 'member-model' into member-password-reset

Conflicts:
	app/controllers/members_controller.rb
	config/routes.rb
	spec/controllers/members_controller_spec.rb

nixpulvis committed:
Don't use attr_protected.

This was causing problems because it was requiring,
every model to follow the old style of attr protection.
We'll use the new rails 4 method and move this logic in
to the controllers.

nixpulvis committed:
Don't use a before filter to assign @member.

This was a good idea at first. But as we add actions that do lookups by
email this was becoming a pain in the ass.

nixpulvis committed:
Adding the method for updating password.

`update_password` will update the member's password,
and if a :key if given it will validate the key.

nixpulvis committed:
Clean up logic for forbidding keys.

nixpulvis commented:
All of this work will be redone with [Devise](https://github.com/plataformatec/devise) and [Devise LDAP Authenticatable](https://github.com/cschiewek/devise_ldap_authenticatable)
